rfcs/45-stageless.md

```diff
@@ -0,0 +1,725 @@
+   // as the identifier would not be stable when systems were removed, risking serious bugs
+   systems: HashMap<SystemId, ConfiguredSystem>,
+   // Stores the fully initialized, efficient graph of ordering constraints
+   // for all systems in the schedule.
+   // The properties of each set are passed down to specific systems to cache computation results.
+   // `Graph` is a bikeshed-avoidance graph storage structure,
+   // the exact strategy should be benchmarked
+   ordering_constraints: Graph<SystemId>,
+   // Set configuration must be stored at the schedule level,
+   // rather than attached to specific set instances
+   sets: HashMap<Box<dyn SystemLabel>, SystemSetConfig>,
+   // Used to check if mutation is allowed
+   currently_running: bool,
+}
+```
+
+Configured systems are composed of a raw system and its configuration.
+
+```rust
+struct ConfiguredSystem{
+   raw_system: System,
+   config: SystemConfig,
+}
+```
+
+Raw systems store metadata, cached state, `Local` system parameters and the function pointer to the actual function to be executed:
+
+```rust
+struct System {
+   function: F,
+   // Contains all metadata and state
+   meta: SystemMeta
+   last_change_tick: u32,
+}
+```
+
+System configuration stores ordering dependencies, run criteria and system sets.
+It can be created raw, or stored in association with either system sets or systems.
+
+```rust
+enum OrderingTarget {
+   Label(Box<dyn SystemLabel>),
+   System(SystemId),
+}
+
+struct SystemConfig {
+   strict_ordering_before: Vec<OrderingTarget>,
+   strict_ordering_after: Vec<OrderingTarget>,
+   // The `TypeId` here is the type id of the flushing system
+   flushed_ordering_before: Vec<(TypeId, OrderingTarget)>,
+   flushed_ordering_after: Vec<(TypeId, OrderingTarget)>,
+   run_criteria: Vec<SystemId>,
+   // The combined access requirements of the system and its run criteria
+   joint_access: FilteredAccessSet<ComponentId>,
+   sets: Vec<Box dyn SystemLabel>,
+}
+
+struct SystemSetConfig {
+   system_config: SystemConfig,
+}
+
+```
+
+### System set configuration
+
+When building the dependency graph, all edges (constraints) between labels are expanded into edges between the systems nested within them.
+
+### Run criteria
+
+Run criteria are systems with several key constraints:
+
+- They cannot have system params that give mutable access to data.
+  - Required so we don't have order to them.
+  - Read-only access avoids very surprising side effects.
+- They cannot capture or store local data.
+  - Required to avoid user footguns: local data would be updated even if the guarded system ultimately doesn't run.
+- They must return a `bool`.
+  - Required so we know whether the system should run!
+
+Run criteria are shared.
+They are evaluated a single time during each schedule pass, and the result applies to all systems with that run criteria for the tick.
+
+ When the executor is presented with a ready system, we do the following:

```

Comment:
What does it mean for a system to be ready?

Reply:
Ah, good catch. "Ready" means that all its dependencies (whatever the system was configured to run after) have completed.

Reply:
Cool, that definitely needs to be defined in the RFC.

Even the terminology for “ordering dependency“ is vaguely defined, but it can be quite easily be understood. A more formal definition though would remove any potential conceptual ambiguity.

I thought about these definitions (all the following definitions are only referring to the context of a single schedule run, not to multiple runs, adjacent ones in particular, and they are not referring to the entire schedule lifecycle):

> Let `A` and `B` be systems. If `B` is set to run after `A`, then `B` is said to have a **dependency** on `A`.
>
> A system is said to have **completed** from the moment it has run or skipped.
>
> A system is said to be **ready** if all of its dependencies have completed.

I admit that my definition of completed may a little bit flimsy, as it could be expressed more concisely by introducing the executor. But that would have a couple of problems:

1. There's a risk of delving to much into details, while we want to give an overview.
2. The “completed” definition would boil down to just “the executor marked the system as such”, which is not very helpful.

At the end, I think that the executor-less definition is better for an introductory definitions section. We can still define the executor later, but it looks more of an implementation detail than other things (and as such it could be defined in the implementation strategy).